File: clients/android/app/src/main/java/network/buildit/modules/search/domain/TFIDFEngine.kt
package network.buildit.modules.search.domain

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import network.buildit.modules.search.data.InverseDocumentFrequencyDao
import network.buildit.modules.search.data.InverseDocumentFrequencyEntity
import network.buildit.modules.search.data.SearchDocumentDao
import network.buildit.modules.search.data.TermFrequencyDao
import network.buildit.modules.search.data.TermFrequencyEntity
import network.buildit.modules.search.models.SearchDocument
import network.buildit.modules.search.models.SparseVector
import java.text.Normalizer
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.ln
import kotlin.math.sqrt

/**
 * TF-IDF (Term Frequency-Inverse Document Frequency) engine for semantic ranking.
 *
 * Provides:
 * - Text tokenization and normalization
 * - TF-IDF vector computation
 * - Cosine similarity scoring
 * - Incremental index updates
 *
 * The engine uses a sparse vector representation where only non-zero terms are stored,
 * making it memory-efficient for large vocabularies.
 */
@Singleton
class TFIDFEngine @Inject constructor(
    private val searchDocumentDao: SearchDocumentDao,
    private val termFrequencyDao: TermFrequencyDao,
    private val idfDao: InverseDocumentFrequencyDao
) {
    companion object {
        // Common English stop words to exclude from indexing
        private val STOP_WORDS = setOf(
            "a", "an", "and", "are", "as", "at", "be", "by", "for", "from",
            "has", "he", "in", "is", "it", "its", "of", "on", "that", "the",
            "to", "was", "were", "will", "with", "the", "this", "but", "they",
            "have", "had", "what", "when", "where", "who", "which", "why", "how",
            "all", "each", "every", "both", "few", "more", "most", "other",
            "some", "such", "no", "nor", "not", "only", "own", "same", "so",
            "than", "too", "very", "just", "can", "should", "now", "been", "being"
        )

        // Minimum term length to index
        private const val MIN_TERM_LENGTH = 2

        // Maximum term length to index
        private const val MAX_TERM_LENGTH = 50

        // Smoothing factor for IDF calculation (prevents division by zero)
        private const val IDF_SMOOTHING = 1.0
    }

    /**
     * Tokenizes and normalizes text into indexable terms.
     *
     * Applies:
     * - Unicode normalization (NFD)
     * - Lowercasing
     * - Punctuation removal
     * - Stop word filtering
     * - Length filtering
     * - Basic stemming (suffix stripping)
     */
    fun tokenize(text: String): List<String> {
        // Normalize unicode
        val normalized = Normalizer.normalize(text, Normalizer.Form.NFD)
            .replace(Regex("\\p{M}"), "") // Remove diacritics

        // Split into words, lowercase, and filter
        return normalized
            .lowercase(Locale.ROOT)
            .split(Regex("[\\s\\p{Punct}]+"))
            .filter { token ->
                token.length in MIN_TERM_LENGTH..MAX_TERM_LENGTH &&
                        token !in STOP_WORDS &&
                        token.any { it.isLetter() } // Must contain at least one letter
            }
            .map { stem(it) }
            .distinct()
    }

    /**
     * Basic Porter-like stemmer for English.
     * Removes common suffixes to normalize word forms.
     */
    private fun stem(word: String): String {
        var result = word

        // Remove common suffixes
        val suffixes = listOf(
            "ingly", "ingly", "ation", "ments", "ment", "ness", "less",
            "able", "ible", "ful", "ous", "ive", "ing", "ies", "ied",
            "ion", "ity", "ers", "est", "ed", "ly", "er", "es", "s"
        )

        for (suffix in suffixes) {
            if (result.endsWith(suffix) && result.length - suffix.length >= 3) {
                result = result.dropLast(suffix.length)
                break
            }
        }

        return result
    }

    /**
     * Computes term frequencies for a document.
     *
     * @param text The document text
     * @return Map of term to raw frequency
     */
    fun computeTermFrequencies(text: String): Map<String, Int> {
        val tokens = tokenize(text)
        return tokens.groupingBy { it }.eachCount()
    }

    /**
     * Computes normalized term frequency using augmented frequency.
     * Prevents bias toward longer documents.
     *
     * tf(t,d) = 0.5 + 0.5 * (f(t,d) / max{f(t',d)})
     */
    fun computeNormalizedTf(termFrequencies: Map<String, Int>): Map<String, Double> {
        if (termFrequencies.isEmpty()) return emptyMap()

        val maxFreq = termFrequencies.values.maxOrNull()?.toDouble() ?: 1.0
        return termFrequencies.mapValues { (_, freq) ->
            0.5 + 0.5 * (freq / maxFreq)
        }
    }

    /**
     * Computes IDF value for a term.
     *
     * idf(t) = log(N / (df(t) + 1)) + 1
     *
     * Where N is total documents and df(t) is documents containing term t.
     * The +1 smoothing prevents log(0) and ensures rare terms get higher weight.
     */
    fun computeIdf(totalDocuments: Int, documentFrequency: Int): Double {
        return ln((totalDocuments.toDouble() + IDF_SMOOTHING) / (documentFrequency + IDF_SMOOTHING)) + 1.0
    }

    /**
     * Computes TF-IDF vector for a document.
     */
    suspend fun computeTfIdfVector(
        documentId: String,
        text: String,
        groupId: String
    ): SparseVector = withContext(Dispatchers.Default) {
        val termFreqs = computeTermFrequencies(text)
        val normalizedTf = computeNormalizedTf(termFreqs)

        // Get IDF values for terms
        val terms = termFreqs.keys.toList()
        val idfEntities = idfDao.getIdfsForTerms(terms)
        val idfMap = idfEntities.associate { it.term to it.idfValue }

        // Compute TF-IDF for each term
        val tfidfVector = mutableMapOf<String, Double>()
        for ((term, tf) in normalizedTf) {
            val idf = idfMap[term] ?: 1.0 // Default IDF for new terms
            tfidfVector[term] = tf * idf
        }

        // L2 normalize the vector
        normalizeVector(tfidfVector)
    }

    /**
     * L2 normalizes a vector in place.
     */
    private fun normalizeVector(vector: MutableMap<String, Double>): Map<String, Double> {
        val magnitude = sqrt(vector.values.sumOf { it * it })
        if (magnitude > 0) {
            for ((key, value) in vector) {
                vector[key] = value / magnitude
            }
        }
        return vector
    }

    /**
     * Computes cosine similarity between two TF-IDF vectors.
     * Returns value between 0 (no similarity) and 1 (identical).
     */
    fun cosineSimilarity(vec1: SparseVector, vec2: SparseVector): Double {
        if (vec1.isEmpty() || vec2.isEmpty()) return 0.0

        // Compute dot product
        var dotProduct = 0.0
        for ((term, weight1) in vec1) {
            val weight2 = vec2[term] ?: continue
            dotProduct += weight1 * weight2
        }

        // Vectors are already normalized, so dot product is the cosine similarity
        return dotProduct
    }

    /**
     * Indexes a document for TF-IDF search.
     *
     * @param document The document to index
     * @return The computed TF-IDF vector
     */
    suspend fun indexDocument(document: SearchDocument): SparseVector = withContext(Dispatchers.IO) {
        // Combine searchable text
        val fullText = buildString {
            append(document.title)
            append(" ")
            append(document.content)
            append(" ")
            document.tags.forEach { tag ->
                append(tag)
                append(" ")
            }
        }

        // Compute term frequencies
        val termFreqs = computeTermFrequencies(fullText)
        val normalizedTf = computeNormalizedTf(termFreqs)

        // Store term frequencies
        val termFrequencyEntities = normalizedTf.map { (term, normalizedFreq) ->
            TermFrequencyEntity(
                documentId = document.id,
                term = term,
                groupId = document.groupId,
                frequency = termFreqs[term] ?: 0,
                normalizedFrequency = normalizedFreq
            )
        }

        // Delete existing term frequencies for this document
        termFrequencyDao.deleteTermsForDocument(document.id)
        termFrequencyDao.insertTermFrequencies(termFrequencyEntities)

        // Update IDF values for affected terms
        updateIdfForTerms(termFreqs.keys.toList())

        // Compute and return TF-IDF vector
        computeTfIdfVector(document.id, fullText, document.groupId)
    }

    /**
     * Updates IDF values for a set of terms.
     */
    suspend fun updateIdfForTerms(terms: List<String>) = withContext(Dispatchers.IO) {
        val totalDocuments = searchDocumentDao.getDocumentCount()
        val currentTime = System.currentTimeMillis()

        val idfEntities = terms.map { term ->
            val documentFreq = termFrequencyDao.getDocumentsForTerm(term).size
            InverseDocumentFrequencyEntity(
                term = term,
                documentCount = documentFreq,
                idfValue = computeIdf(totalDocuments, documentFreq),
                lastUpdated = currentTime
            )
        }

        idfDao.insertIdfs(idfEntities)
    }

    /**
     * Removes a document from the TF-IDF index.
     */
    suspend fun removeDocument(documentId: String) = withContext(Dispatchers.IO) {
        // Get terms to update before deleting
        val terms = termFrequencyDao.getTermsForDocument(documentId)
        val termStrings = terms.map { it.term }

        // Delete term frequencies
        termFrequencyDao.deleteTermsForDocument(documentId)

        // Update IDF values for affected terms
        updateIdfForTerms(termStrings)
    }

    /**
     * Performs semantic search using TF-IDF vectors.
     *
     * @param query The search query
     * @param candidateDocuments Documents to score (pre-filtered by FTS)
     * @param limit Maximum results to return
     * @return List of document IDs with similarity scores, sorted by score descending
     */
    suspend fun semanticSearch(
        query: String,
        candidateDocuments: List<SearchDocument>,
        limit: Int = 50
    ): List<Pair<String, Double>> = withContext(Dispatchers.Default) {
        if (candidateDocuments.isEmpty()) return@withContext emptyList()

        // Compute query vector
        val queryTermFreqs = computeTermFrequencies(query)
        val normalizedQueryTf = computeNormalizedTf(queryTermFreqs)

        // Get IDF values for query terms
        val queryTerms = queryTermFreqs.keys.toList()
        val idfEntities = idfDao.getIdfsForTerms(queryTerms)
        val idfMap = idfEntities.associate { it.term to it.idfValue }

        // Compute query TF-IDF vector
        val queryVector = mutableMapOf<String, Double>()
        for ((term, tf) in normalizedQueryTf) {
            val idf = idfMap[term] ?: 1.0
            queryVector[term] = tf * idf
        }
        normalizeVector(queryVector)

        // Score each candidate document
        val scores = candidateDocuments.mapNotNull { doc ->
            val docVector = doc.vector ?: return@mapNotNull null
            val similarity = cosineSimilarity(queryVector, docVector)
            if (similarity > 0) {
                doc.id to similarity
            } else {
                null
            }
        }

        // Sort by similarity and return top results
        scores.sortedByDescending { it.second }.take(limit)
    }

    /**
     * Rebuilds the entire TF-IDF index.
     * Should be called periodically or after significant changes.
     */
    suspend fun rebuildIndex() = withContext(Dispatchers.IO) {
        // Clear existing index
        termFrequencyDao.deleteAllTermFrequencies()
        idfDao.deleteAllIdfs()

        // Get all documents in batches
        var offset = 0
        val batchSize = 100
        val allTerms = mutableSetOf<String>()

        while (true) {
            val documents = searchDocumentDao.getAllDocuments(batchSize, offset)
            if (documents.isEmpty()) break

            for (entity in documents) {
                val document = entity.toSearchDocument()
                val fullText = buildString {
                    append(document.title)
                    append(" ")
                    append(document.content)
                    append(" ")
                    document.tags.forEach { tag ->
                        append(tag)
                        append(" ")
                    }
                }

                val termFreqs = computeTermFrequencies(fullText)
                val normalizedTf = computeNormalizedTf(termFreqs)

                val termFrequencyEntities = normalizedTf.map { (term, normalizedFreq) ->
                    TermFrequencyEntity(
                        documentId = document.id,
                        term = term,
                        groupId = document.groupId,
                        frequency = termFreqs[term] ?: 0,
                        normalizedFrequency = normalizedFreq
                    )
                }

                termFrequencyDao.insertTermFrequencies(termFrequencyEntities)
                allTerms.addAll(termFreqs.keys)
            }

            offset += batchSize
        }

        // Update all IDF values
        updateIdfForTerms(allTerms.toList())
    }

    /**
     * Gets index statistics.
     */
    suspend fun getIndexStats(): TFIDFStats = withContext(Dispatchers.IO) {
        TFIDFStats(
            uniqueTerms = termFrequencyDao.getUniqueTermCount(),
            totalDocuments = searchDocumentDao.getDocumentCount()
        )
    }
}

/**
 * Statistics about the TF-IDF index.
 */
data class TFIDFStats(
    val uniqueTerms: Int,
    val totalDocuments: Int
)

File: clients/web/src/modules/search/services/tfidfEngine.ts
/**
 * TF-IDF Engine Service
 * Provides semantic search capabilities using term frequency-inverse document frequency
 */

import type { SearchDocument, SparseVector, DocumentStats, TermFrequency } from '../types';

// ============================================================================
// Constants
// ============================================================================

/**
 * Stop words to exclude from TF-IDF calculations
 */
const STOP_WORDS = new Set([
  'a', 'an', 'and', 'are', 'as', 'at', 'be', 'by', 'for', 'from',
  'has', 'he', 'in', 'is', 'it', 'its', 'of', 'on', 'or', 'she',
  'that', 'the', 'they', 'to', 'was', 'were', 'which', 'with',
  'you', 'your', 'this', 'these', 'those', 'i', 'me', 'my', 'we',
  'our', 'will', 'would', 'could', 'should', 'can', 'may', 'have',
  'had', 'do', 'does', 'did', 'been', 'being', 'am', 'if', 'so',
  'but', 'not', 'no', 'than', 'then', 'there', 'here', 'when',
  'where', 'what', 'who', 'how', 'why', 'all', 'each', 'every',
  'both', 'few', 'more', 'most', 'other', 'some', 'such', 'only',
]);

/**
 * Minimum term length to include
 */
const MIN_TERM_LENGTH = 2;

/**
 * Maximum vocabulary size
 */
const MAX_VOCABULARY_SIZE = 50000;

// ============================================================================
// TF-IDF Engine Class
// ============================================================================

export class TFIDFEngine {
  private vocabulary: Map<string, number> = new Map();  // term -> index
  private reverseVocab: Map<number, string> = new Map(); // index -> term
  private documentFrequency: Map<number, number> = new Map();  // term index -> doc count
  private totalDocuments: number = 0;
  private totalTermsSum: number = 0;
  private nextTermIndex: number = 0;

  /**
   * Get vocabulary size
   */
  get vocabularySize(): number {
    return this.vocabulary.size;
  }

  /**
   * Get total documents indexed
   */
  get documentCount(): number {
    return this.totalDocuments;
  }

  /**
   * Get document statistics
   */
  getStats(): DocumentStats {
    const df: Record<string, number> = {};
    for (const [idx, count] of this.documentFrequency.entries()) {
      const term = this.reverseVocab.get(idx);
      if (term) {
        df[term] = count;
      }
    }

    return {
      totalDocuments: this.totalDocuments,
      documentFrequency: df,
      avgDocLength: this.totalDocuments > 0 ? this.totalTermsSum / this.totalDocuments : 0,
    };
  }

  /**
   * Tokenize text into terms
   */
  tokenize(text: string): string[] {
    if (!text) return [];

    return text
      .toLowerCase()
      // Split on non-alphanumeric characters
      .split(/[^a-z0-9]+/)
      // Filter out stop words and short terms
      .filter((term) => term.length >= MIN_TERM_LENGTH && !STOP_WORDS.has(term))
      // Basic stemming
      .map((term) => this.stem(term));
  }

  /**
   * Basic Porter-like stemmer
   */
  private stem(word: string): string {
    // Remove common suffixes
    return word
      .replace(/ies$/, 'i')
      .replace(/es$/, '')
      .replace(/s$/, '')
      .replace(/ing$/, '')
      .replace(/ed$/, '')
      .replace(/ly$/, '')
      .replace(/ment$/, '')
      .replace(/tion$/, 't')
      .replace(/ness$/, '');
  }

  /**
   * Get or create term index
   */
  private getTermIndex(term: string, create: boolean = false): number | undefined {
    let index = this.vocabulary.get(term);

    if (index === undefined && create) {
      if (this.vocabulary.size >= MAX_VOCABULARY_SIZE) {
        // Vocabulary is full, don't add new terms
        return undefined;
      }
      index = this.nextTermIndex++;
      this.vocabulary.set(term, index);
      this.reverseVocab.set(index, term);
    }

    return index;
  }

  /**
   * Compute term frequencies for a document
   */
  computeTermFrequencies(text: string): TermFrequency[] {
    const tokens = this.tokenize(text);
    const counts: Map<string, number> = new Map();

    for (const token of tokens) {
      counts.set(token, (counts.get(token) || 0) + 1);
    }

    const totalTerms = tokens.length;
    const frequencies: TermFrequency[] = [];

    for (const [term, count] of counts.entries()) {
      frequencies.push({
        term,
        count,
        frequency: count / totalTerms,
      });
    }

    return frequencies;
  }

  /**
   * Build TF-IDF vector for a document
   */
  buildVector(doc: SearchDocument): SparseVector {
    // Combine all searchable text
    const text = [
      doc.title,
      doc.title,  // Double-weight title
      doc.content,
      doc.tags.join(' '),
    ].join(' ');

    const tokens = this.tokenize(text);
    const termCounts: Map<number, number> = new Map();
    const termsInDoc: Set<number> = new Set();

    // Count term occurrences
    for (const token of tokens) {
      const idx = this.getTermIndex(token, true);
      if (idx !== undefined) {
        termCounts.set(idx, (termCounts.get(idx) || 0) + 1);
        termsInDoc.add(idx);
      }
    }

    const totalTerms = tokens.length;
    const vector: SparseVector = {};

    // Compute TF-IDF weights
    for (const [termIdx, count] of termCounts.entries()) {
      // Term frequency: count / total terms in document
      const tf = count / totalTerms;

      // Inverse document frequency: log(N / (df + 1))
      const df = this.documentFrequency.get(termIdx) || 0;
      const idf = Math.log((this.totalDocuments + 1) / (df + 1)) + 1;

      // TF-IDF score
      vector[termIdx] = tf * idf;
    }

    return vector;
  }

  /**
   * Add a document to the corpus (updates document frequencies)
   */
  addDocument(doc: SearchDocument): SparseVector {
    const vector = this.buildVector(doc);

    // Update document frequencies
    for (const termIdx of Object.keys(vector).map(Number)) {
      this.documentFrequency.set(
        termIdx,
        (this.documentFrequency.get(termIdx) || 0) + 1
      );
    }

    this.totalDocuments++;
    this.totalTermsSum += Object.keys(vector).length;

    return vector;
  }

  /**
   * Remove a document from the corpus
   */
  removeDocument(vector: SparseVector): void {
    if (this.totalDocuments === 0) return;

    // Decrement document frequencies
    for (const termIdx of Object.keys(vector).map(Number)) {
      const currentDf = this.documentFrequency.get(termIdx) || 0;
      if (currentDf > 1) {
        this.documentFrequency.set(termIdx, currentDf - 1);
      } else {
        this.documentFrequency.delete(termIdx);
      }
    }

    this.totalDocuments--;
    this.totalTermsSum -= Object.keys(vector).length;
  }

  /**
   * Compute TF-IDF vector for a query
   */
  buildQueryVector(query: string): SparseVector {
    const tokens = this.tokenize(query);
    const termCounts: Map<number, number> = new Map();

    for (const token of tokens) {
      // Don't create new terms for query - only match existing vocabulary
      const idx = this.getTermIndex(token, false);
      if (idx !== undefined) {
        termCounts.set(idx, (termCounts.get(idx) || 0) + 1);
      }
    }

    const totalTerms = tokens.length || 1;
    const vector: SparseVector = {};

    for (const [termIdx, count] of termCounts.entries()) {
      const tf = count / totalTerms;
      const df = this.documentFrequency.get(termIdx) || 0;
      const idf = Math.log((this.totalDocuments + 1) / (df + 1)) + 1;
      vector[termIdx] = tf * idf;
    }

    return vector;
  }

  /**
   * Compute cosine similarity between two sparse vectors
   */
  cosineSimilarity(vecA: SparseVector, vecB: SparseVector): number {
    // Dot product
    let dotProduct = 0;
    for (const [idx, weightA] of Object.entries(vecA)) {
      const weightB = vecB[Number(idx)];
      if (weightB !== undefined) {
        dotProduct += weightA * weightB;
      }
    }

    // Magnitudes
    let magA = 0;
    for (const weight of Object.values(vecA)) {
      magA += weight * weight;
    }
    magA = Math.sqrt(magA);

    let magB = 0;
    for (const weight of Object.values(vecB)) {
      magB += weight * weight;
    }
    magB = Math.sqrt(magB);

    if (magA === 0 || magB === 0) return 0;

    return dotProduct / (magA * magB);
  }

  /**
   * Find similar documents using TF-IDF vectors
   */
  findSimilar(
    queryVector: SparseVector,
    documentVectors: Array<{ id: string; vector: SparseVector }>,
    limit: number = 10,
    minSimilarity: number = 0.1
  ): Array<{ id: string; similarity: number }> {
    const results: Array<{ id: string; similarity: number }> = [];

    for (const { id, vector } of documentVectors) {
      const similarity = this.cosineSimilarity(queryVector, vector);
      if (similarity >= minSimilarity) {
        results.push({ id, similarity });
      }
    }

    // Sort by similarity descending
    results.sort((a, b) => b.similarity - a.similarity);

    return results.slice(0, limit);
  }

  /**
   * Get top terms from a vector
   */
  getTopTerms(vector: SparseVector, limit: number = 10): Array<{ term: string; weight: number }> {
    const terms: Array<{ term: string; weight: number }> = [];

    for (const [idx, weight] of Object.entries(vector)) {
      const term = this.reverseVocab.get(Number(idx));
      if (term) {
        terms.push({ term, weight });
      }
    }

    terms.sort((a, b) => b.weight - a.weight);
    return terms.slice(0, limit);
  }

  /**
   * Clear the engine state
   */
  clear(): void {
    this.vocabulary.clear();
    this.reverseVocab.clear();
    this.documentFrequency.clear();
    this.totalDocuments = 0;
    this.totalTermsSum = 0;
    this.nextTermIndex = 0;
  }

  /**
   * Export engine state for persistence
   */
  exportState(): string {
    return JSON.stringify({
      vocabulary: Array.from(this.vocabulary.entries()),
      documentFrequency: Array.from(this.documentFrequency.entries()),
      totalDocuments: this.totalDocuments,
      totalTermsSum: this.totalTermsSum,
      nextTermIndex: this.nextTermIndex,
    });
  }

  /**
   * Import engine state
   */
  importState(json: string): void {
    const data = JSON.parse(json);

    this.vocabulary = new Map(data.vocabulary);
    this.reverseVocab = new Map(
      data.vocabulary.map(([term, idx]: [string, number]) => [idx, term])
    );
    this.documentFrequency = new Map(data.documentFrequency);
    this.totalDocuments = data.totalDocuments;
    this.totalTermsSum = data.totalTermsSum;
    this.nextTermIndex = data.nextTermIndex;
  }
}

// ============================================================================
// Singleton Instance
// ============================================================================

let instance: TFIDFEngine | null = null;

/**
 * Get the singleton TF-IDF engine instance
 */
export function getTFIDFEngine(): TFIDFEngine {
  if (!instance) {
    instance = new TFIDFEngine();
  }
  return instance;
}

/**
 * Reset the singleton instance (for testing)
 */
export function resetTFIDFEngine(): void {
  if (instance) {
    instance.clear();
  }
  instance = null;
}
